runtime/near-vm-runner/RUNTIMES.md

```diff
@@ -0,0 +1,176 @@
+matters.
+
+On a scale of trade-offs, runtime performance is probably one of the less important metrics. Slower
+execution of a contract doesn't make NEAR protocol unsound as an idea, whereas something like a
+non-deterministic execution outcome would.
+
+## Preparation performance
+
+The NEAR protocol charges gas fees to deploy a contract. One part of a deployment involves
+preparing the contract Wasm code for execution. This may involve instrumentation, compilation,
+serialization of the machine code and other similar operations.
+
+The gas costs are charged before any work is done towards the deployment, so the cost of the
+preparation, compilation and similar tasks must be predictable with a comparatively high precision.
+
+An estimate of the deployment cost will typically involve a function which uses the size of the
+input Wasm code as its primary input. Such a function can only exist if we have a good knowledge of
+the runtime’s time complexity properties are known. For our purposes a linear or `O(n log n)`
+relationship between the input size and execution time is the highest we can accept.
+
+# Evaluation
+
+We have evaluated a number of different implementations over time. Currently Wasmer with the
+universal engine and the single-pass backend is the default.
+
+## `wasmer-singlepass` codegen
+
+The `wasmer-singlepass` code generator is an implementation of a straightforward, direct
+translation from Wasm statements to machine code. This approach is linear over the input size by
+construction satisfying the requirements for preparation performance.
+
+A single-pass codegen implementation in general incurs a trade-off in runtime performance.
+WebAssembly is modelling a stack machine and an obviously correct implementation of such a machine
+would be to directly translate all the stack operations to native code as well. Targets running
+this code, on the other hand, are most likely an implementation of register machine where stack
+operations are not necessarily all that efficient.
+
+In order to improve the quality of the generated machine code, `wasmer-singlepass` maintains some
+state between the individual Wasm instructions. This state allows `wasmer-singlepass` to avoid some
+stack operations, pick more efficient instructions and sometimes avoid some operations such as NaN
+canonicalization altogether.
+
+Presence of this state comes at a cost of testability of implementation. The behaviour of the
+codegen can depend on prior Wasm instructions. Because of this there's a combinatorial explosion of
+the test cases that need to be considered and verified before confidence in the results can be
+established. Relatedly, at times it may also be difficult to validate changes made to the
+implementation of the `wasmer-singlepass` codegen. The global state is definitely a source of
+potential spooky action at a distance problems where changing code generation of a specific
+instruction affects correctness and behaviour of another one.
```

Comment:
Curious, is this theoretical, or was 2.1.0 miscompilation an example of that?

Reply:
2.1.0 calling convention miscompile is not strictly an example of the issue per se. It was a problem with how calling convention was handled and the issue was introduced by changes to the area of code that was dealing with calling conventions.

On the other hand, it is in some sense such an example. In order to reproduce this issue a number of other instructions preceding the miscompiled call were necessary. These instructions seem to have been responsible for setting up the global state just the right way to trigger the invalid case in the code for call codegen.